### PR TITLE
Controlplane: start watching stop polling

### PR DESCRIPTION
**Pull request by aojea**
/kind bug
/kind cleanup

The apiserver runs a bootstrap controller that starts 3 reconcile loops:

- kubernetes.default services and endpoints reconciler (if configured)
- system namespaces
- services ClusterIP and NodePort repair loops

These reconcile loops runs every certain period performing API request, in each loop.

This PR split the bootstrap controller in 3 independent controllers that use informers, and start using watchers/listers for each controller, improving its performance.

```release-note
The apiserver bootstrap controller is split into 3 different controllers, the healthz/readyz endpoints reflect these changes
		"[+]poststarthook/bootstrap-controller ok",
		"[+]poststarthook/start-default-service-controller ok",
		"[+]poststarthook/start-system-namespaces-controller ok",
		"[+]poststarthook/start-services-repair-controller ok",
The new controllers improve the apiserver performance, using watchers instead of polling.
The `start-default-service-controller`, that reconciles the `kubernetes.default` Service pointing to the apiserver, reconciles the `kubernetes.default` Service state, updating or deleting the Service accordingly to match the configured parameters. Previously it required manual intervention if the Service was modified or deleted.
```

### CODE REVIEW

File: pkg/controlplane/controller.go [hunk 1]
```diff
@@ -236,141 +163,3 @@ func (c *Controller) RunKubernetesNamespaces(ch chan struct{}) {
...
-		var code int
-		c.client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do(context.TODO()).StatusCode(&code)
-		return code == http.StatusOK, nil
-	}, ch)
-
-	wait.NonSlidingUntil(func() {
-		// Service definition is not reconciled after first
-		// run, ports and type will be corrected only during
-		// start.
-		if err := c.UpdateKubernetesService(false); err != nil {
```
Comment (aojea) [inline pkg/controlplane/controller.go:253]: actually Service definition seems is never reconciled since https://github.com/kubernetes/kubernetes/commit/2a9a9fa155e92d0bd7f4eb85d843cfa14b830cca

File: pkg/controlplane/controller.go [hunk 2]
```diff
@@ -236,141 +163,3 @@ func (c *Controller) RunKubernetesNamespaces(ch chan struct{}) {
...
-		Name:     endpointPortName,
-	}}
-}
-
-// CreateOrUpdateMasterServiceIfNeeded will create the specified service if it
-// doesn't already exist.
-func (c *Controller) CreateOrUpdateMasterServiceIfNeeded(serviceName string, serviceIP net.IP, servicePorts []corev1.ServicePort, serviceType corev1.ServiceType, reconcile bool) error {
-	if s, err := c.client.CoreV1().Services(metav1.NamespaceDefault).Get(context.TODO(), serviceName, metav1.GetOptions{}); err == nil {
-		// The service already exists.
-		if reconcile {
```
Comment (aojea) [inline pkg/controlplane/controller.go:313]: this seems to be always false

**Comment by aojea** [comment]
/assign @thockin @liggitt @wojtek-t @deads2k 

This is required to implement the KEP https://github.com/kubernetes/kubernetes/pull/110470, but when doing this refactor I found out that current behavior is not what I expected, the kubernetes. default Service  doesn't seem to be reconciled , and this is since https://github.com/kubernetes/kubernetes/pull/74668 1.14.

This means that current behavior is "First Win", multiple apiservers with different ClusterIP for the apiserver configured (different ServiceIPSubnet), will not try to reconcile the Service, however, the Service repair loop will break the Services as it will rewrite the allocator bitmap with different snapshots ...

TL;DR 
- multiple apiserver with different configured service-ipranges does not work and never worked
- if an admin modifies the kubernetes.default service, this is never reconciled

With the new KEP, we fix the multiple service-cidr configuration problem, implementing the same conflict resolution, the "First apiserver that creates kubernetes.default" wins.

My question is, should I fix the reconciliation of the kubernetes.default service for anything that is not the ClusterIP?

File: pkg/controlplane/controller.go [hunk 3]
```diff
@@ -17,41 +17,33 @@ limitations under the License.
...
 const (
 	kubernetesServiceName = "kubernetes"
 )
 
 // Controller is the controller manager for the core bootstrap Kubernetes
-// controller loops, which manage creating the "kubernetes" service, the
-// "default", "kube-system" and "kube-public" namespaces, and provide the IP
-// repair check on service IPs
+// controller loops, the "default", "kube-system" and "kube-public" namespaces,
+// and provide the IP repair check on service IPs
```
Comment (aojea) [inline pkg/controlplane/controller.go:46]: I'm going to split each of them into its own controllers, with KEP-1880 Multiple Service CIDRs the ClusterIP repairs loop is moved to the controller manager (but I also think we can move both repair loops to the controller manager 
 without that KEP)

File: pkg/controlplane/controller.go [hunk 4]
```diff
@@ -236,141 +163,3 @@ func (c *Controller) RunKubernetesNamespaces(ch chan struct{}) {
...
-		}
-	}, c.EndpointInterval, ch)
-}
-
-// UpdateKubernetesService attempts to update the default Kube service.
-func (c *Controller) UpdateKubernetesService(reconcile bool) error {
-	// Update service & endpoint records.
-	// TODO: when it becomes possible to change this stuff,
-	// stop polling and start watching.
-	// TODO: add endpoints of all replicas, not just the elected master.
```
Comment (aojea) [inline pkg/controlplane/controller.go:264]: this TODO are addressed with this PR, the endpoints TODO is already handled by the endpoints reconciler

**Comment by aojea** [comment]
/retest

Errors don't look related

**Comment by sftim** [comment]
For the changelog, please consider splitting this into a brief summary and an extra details block.

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// Reconcile during first run removing itself until server is ready.
+	endpointPorts := createEndpointPortSpec(c.publicServicePort, "https")
+	if err := c.endpointReconciler.RemoveEndpoints(serviceName, c.publicIP, endpointPorts); err == nil {
+		klog.Error("Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously")
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:148]:
```suggestion
		klog.Error("Found stale data, removed previous endpoints for \"kubernetes\" Service. API server didn't exit successfully previously")
```
?

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+	} else if !storage.IsNotFound(err) {
+		klog.Errorf("Error removing old endpoints from kubernetes service: %v", err)
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:150]:
```suggestion
		klog.Errorf("Error removing old endpoints from \"kubernetes\" Service: %v", err)
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.servicesSynced) {
+		return
+	}
+
+	// wait to publish the endpoints until the process is ready so we don't receive traffic
+	wait.PollImmediateUntil(100*time.Millisecond, func() (bool, error) {
+		var code int
+		c.client.CoreV1().RESTClient().Get().AbsPath("/healthz").Do(context.Background()).StatusCode(&code)
+		return code == http.StatusOK, nil
+	}, stopCh)
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:164]: As a future improvement, I suppose we could publish into `notReadyAddresses` whilst pending?

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+
+	wait.NonSlidingUntil(func() {
+		if err := c.sync(); err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to sync kubernetes.default service: %v", err))
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:168]:
```suggestion
			utilruntime.HandleError(fmt.Errorf("unable to sync kubernetes.default Service: %v", err))
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+	c.stopOnce.Do(c.stop)
+}
+
+// Stop cleans up this API Servers endpoint reconciliation leases so another master can take over more quickly.
+func (c *Controller) stop() {
+	endpointPorts := createEndpointPortSpec(c.publicServicePort, "https")
+	finishedReconciling := make(chan struct{})
+	go func() {
+		defer close(finishedReconciling)
+		klog.Infof("Shutting down kubernetes service endpoint reconciler")
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:181]:
```suggestion
		klog.Infof("Shutting down kubernetes Service endpoint reconciler")
```
?

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+		c.endpointReconciler.StopReconciling()
+		if err := c.endpointReconciler.RemoveEndpoints(serviceName, c.publicIP, endpointPorts); err != nil {
+			klog.Errorf("Unable to remove endpoints from kubernetes service: %v", err)
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:184]:
```suggestion
			klog.Errorf("Unable to remove endpoints from kubernetes Service: %v", err)
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+		return err
+	}
+	return nil
+}
+
+// createPortAndServiceSpec creates an array of service ports.
+// If the NodePort value is 0, just the servicePort is used, otherwise, a node port is exposed.
+func createPortAndServiceSpec(servicePort int, targetServicePort int, nodePort int, servicePortName string) ([]v1.ServicePort, v1.ServiceType) {
+	// Use the Cluster IP type for the service port if NodePort isn't provided.
+	// Otherwise, we will be binding the master service to a NodePort.
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:230]:
```suggestion
	// Otherwise, we will be binding the API server service to a NodePort.
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+		},
+	}
+	_, err := c.client.CoreV1().Namespaces().Create(context.TODO(), newNs, metav1.CreateOptions{})
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = nil
+	}
+	return err
+}
+
+// createOrUpdateMasterServiceIfNeeded will create the specified service if it doesn't already exist.
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:272]:
```suggestion
// createOrUpdateMasterServiceIfNeeded will create the specified service if it doesn't already exist.
// TODO: rename to avoid use of term "master"
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+		_, err := c.client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+		return err
+	}
+	// unexpected error
+	if err != nil {
+		return err
+	}
+	// Service exist, reconcile if necessar
+	if serviceNeedsUpdate(svc, s) {
+		klog.Warningf("Resetting master service %q to %#v", serviceName, svc)
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:305]:
```suggestion
	// Service exists; reconcile if necessary
	if serviceNeedsUpdate(svc, s) {
		klog.Warningf("Resetting API server Service %q to %#v", serviceName, svc)
```

File: pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+	go wait.Until(c.sync, c.interval, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) sync() {
+	// Loop the system namespace list, and create them if they do not exist
+	for _, ns := range c.systemNamespaces {
+		if err := c.createNamespaceIfNeeded(ns); err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to create required kubernetes system namespace %s: %v", ns, err))
```
Comment (sftim) [inline pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go:83]:
```suggestion
			utilruntime.HandleError(fmt.Errorf("unable to create required kubernetes system Namespace %s: %v", ns, err))
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+	client clientset.Interface,
+) (*Controller, error) {
+	// The "kubernetes.default" Service is SingleStack based on the configured ServiceIPRange.
+	// If the bootstrap controller reconcile the kubernetes.default Service and Endpoints, it must
+	// guarantee that the Service ClusterIP and the associated Endpoints have the same IP family, or
+	// it will not work for clients because of the IP family mismatch.
+	// TODO: revisit for dual-stack https://github.com/kubernetes/enhancements/issues/2438
+	if endpointReconcilerType != reconcilers.NoneEndpointReconcilerType {
+		if netutils.IsIPv4(serviceIP) != netutils.IsIPv4(publicIP) {
+			return nil, fmt.Errorf("service IP family %q must match endpoint (apiserver) address family %q", serviceIP.String(), publicIP.String())
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:73]:
```suggestion
			return nil, fmt.Errorf("service IP family %q must match endpoint (API server) address family %q", serviceIP.String(), publicIP.String())
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+	stopOnce sync.Once
+}
+
+// Run will not return until ctx.Done() is closed.
+func (c *Controller) start(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	klog.Infof("Sending events to api server.")
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:138]:
```suggestion
	klog.Infof("Sending events to API server.")
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+	wait.NonSlidingUntil(func() {
+		if err := c.sync(); err != nil {
+			utilruntime.HandleError(fmt.Errorf("unable to sync kubernetes.default service: %v", err))
+		}
+	}, c.endpointInterval, stopCh)
+
+	c.stopOnce.Do(c.stop)
+}
+
+// Stop cleans up this API Servers endpoint reconciliation leases so another master can take over more quickly.
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:175]:
```suggestion
// Stop cleans up this API server's Endpoints reconciliation leases, so that another API server can take over more quickly.
```

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+func createEndpointPortSpec(endpointPort int, endpointPortName string) []v1.EndpointPort {
+	return []v1.EndpointPort{{
+		Protocol: v1.ProtocolTCP,
+		Port:     int32(endpointPort),
+		Name:     endpointPortName,
+	}}
+}
+
+func (c *Controller) createNamespaceIfNeeded() error {
+	if _, err := c.client.CoreV1().Namespaces().Get(context.TODO(), metav1.NamespaceDefault, metav1.GetOptions{}); err == nil {
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:255]: Optionally: comment why `TODO` is appropriate here.

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+		// the namespace already exists
+		return nil
+	}
+	newNs := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metav1.NamespaceDefault,
+			Namespace: "",
+		},
+	}
+	_, err := c.client.CoreV1().Namespaces().Create(context.TODO(), newNs, metav1.CreateOptions{})
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:265]: Optionally: comment why `TODO` is appropriate here.

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+			IPFamilyPolicy:  &singleStack,
+			SessionAffinity: v1.ServiceAffinityNone,
+			Type:            serviceType,
+		},
+	}
+
+	s, err := c.serviceLister.Services(metav1.NamespaceDefault).Get(serviceName)
+	// Service does not exist
+	if err != nil && errors.IsNotFound(err) {
+		_, err := c.client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:296]: Optionally: comment why `TODO` is appropriate here.

File: pkg/controlplane/controller/defaultservice/default_service_controller.go [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+		return err
+	}
+	// unexpected error
+	if err != nil {
+		return err
+	}
+	// Service exist, reconcile if necessar
+	if serviceNeedsUpdate(svc, s) {
+		klog.Warningf("Resetting master service %q to %#v", serviceName, svc)
+		_, err := c.client.CoreV1().Services(metav1.NamespaceDefault).Update(context.TODO(), svc, metav1.UpdateOptions{})
```
Comment (sftim) [inline pkg/controlplane/controller/defaultservice/default_service_controller.go:306]: Optionally: comment why `TODO` is appropriate here.

File: pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+		namespaceSynced:  namespaceInformer.Informer().HasSynced,
+		systemNamespaces: systemNamespaces,
+		interval:         interval,
+	}
+}
+
+// Run starts one worker.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer klog.Infof("Shutting system namespaces controller")
```
Comment (sftim) [inline pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go:65]:
```suggestion
	defer klog.Infof("Shutting down system namespaces controller")
```

File: pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+		// the namespace already exists
+		return nil
+	}
+	newNs := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ns,
+			Namespace: "",
+		},
+	}
+	_, err := c.client.CoreV1().Namespaces().Create(context.TODO(), newNs, metav1.CreateOptions{})
```
Comment (sftim) [inline pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go:99]: Optionally: comment why `TODO` is appropriate here.

**Review by sftim — COMMENTED** [review]
Are these still needed:
```go
	// DefaultEndpointReconcilerInterval is the default amount of time for how often the endpoints for
	// the kubernetes Service are reconciled.
	DefaultEndpointReconcilerInterval = 10 * time.Second
	// DefaultEndpointReconcilerTTL is the default TTL timeout for the storage layer
	DefaultEndpointReconcilerTTL = 15 * time.Second
```
?

---

I have some inline feedback. It might well be that all of those tweaks are either unhelpful or best saved for a follow-up PR. If so, I'm very much fine with that. The suggestions are things that I'd want to see fixed / acked _eventually_.

**Comment by sftim** [comment]
> if an admin modifies the kubernetes.default service, this is never reconciled

If we change that behavior, let's mention it explicitly in the changelog.
Also, we _might_ want to add detail for some circumstances. The one I'm thinking of is if we allow a **delete** and then recreate the deleted Service, I'd want to either or both of:
- send a `Warning:` response for the DELETE (to explain that the API server will try to recreate it)
- create an Event to track that the deletion provokes automatic recreation

**Comment by aojea** [comment]
> Are these still needed:

yeah, `DefaultEndpointReconcilerInterval` is passed as parameter of the `default_service_controller` and `DefaultEndpointReconcilerTTL` of the endpoint reconciler.

The changes keep the same behavior, excepting the `default_service_controller` , because I think that current behaviour is wrong, documented it here https://github.com/kubernetes/kubernetes/issues/114743

**Comment by aojea** [comment]
> If we change that behavior, let's mention it explicitly in the changelog. Also, we _might_ want to add detail for some circumstances. The one I'm thinking of is if we allow a **delete** and then recreate the deleted Service, I'd want to either or both of:
> 
> * send a `Warning:` response for the DELETE (to explain that the API server will try to recreate it)
> * create an Event to track that the deletion provokes automatic recreation

Done, added only the Event by now, I prefer to not touch APIs by now, at least until KEP https://github.com/kubernetes/kubernetes/pull/110470 new APIs are more stable

**Comment by aojea** [comment]
> Kubernetes e2e suite: [It] [sig-storage] PersistentVolumes-local [Volume type: dir-link] Two pods mounting a local volume at the same time should be able to write from pod1 and read from pod2 expand_more

> /test pull-kubernetes-e2e-gce

flaky? since we have enabled flake retries again , @liggitt do you know more about this?

**Comment by aojea** [comment]
> W1230 13:30:38.858] + /go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/ginkgo --nodes=30 --poll-progress-after=300s --poll-progress-interval=20s --source-root=/go/src/k8s.io/kubernetes/kubernetes --timeout=24h --flake-attempts=1 --no-color --output-dir=/workspace/_artifacts --junit-report=ginkgo_report.xml --json-report=ginkgo_report.json /go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/e2e.test -- --kubeconfig=/workspace/.kube/config --host=https://34.82.207.4 --provider=gce --gce-project=k8s-infra-e2e-boskos-112 --gce-zone=us-west1-b --gce-region=us-west1 --gce-multizone=false --gke-cluster=e2e-ecbca2023c-674b9 --kube-master=e2e-ecbca2023c-674b9-master --cluster-tag= --cloud-config-file= --repo-root=/go/src/k8s.io/kubernetes/kubernetes --node-instance-group=e2e-ecbca2023c-674b9-minion-group --prefix=e2e-ecbca2023c-674b9 --network=e2e-ecbca2023c-674b9 --node-tag=e2e-ecbca2023c-674b9-minion --master-tag=e2e-ecbca2023c-674b9-master --docker-config-file= --dns-domain=cluster.local --prepull-images=false --master-os-distro=ubuntu --node-os-distro=ubuntu --num-nodes=3 '--ginkgo.skip=\[Driver:.gcepd\]|\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]' --minStartupPods=8 --report-dir=/workspace/_artifacts --disable-log-dump=true

` --flake-attempts=1` @pohly @onsi does this means it will retry once the test if it fails?

**Comment by onsi** [comment]
no. i know it's confusing but you have to set it to something > 1 to get retries.

**Comment by aojea** [comment]
> no. i know it's confusing but you have to set it to something > 1 to get retries.

Ok, opened #114748 to follow up, I can't understand why it is retrying then

This is good to go for review now

File: pkg/controlplane/controller/defaultservice/default_service_controller.go (hunk 7 above)
Comment (aojea) [reply pkg/controlplane/controller/defaultservice/default_service_controller.go:164]: good point, but we may risk leaking ips and complicate the controller logic to take those into account ... I can't see it is worth the effort just to be more correct

File: pkg/controlplane/controller.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/controlplane/controller.go:253]: Have you followed up on this?  That would be... "bad"?

**Comment by aojea** [comment]
this is not going to happen, "better not touch it if it works"

**Comment by wojtek-t** [comment]
> this is not going to happen, "better not touch it if it works"

Actually - I would like this to happen.
But this PR is huge and I didn't yet find time to review it...
[If we could split it it could potentially help]

**Comment by aojea** [comment]
> Actually - I would like this to happen.

ok, then let me split it in smaller chunks

**Comment by smarterclayton** [comment]
I agree I think this should happen and smaller chunks would help.  I'll commit review time (sorry I didn't get as much in) to it, please copy me in.
